llama/generation.py

```diff
@@ -102,68 +104,124 @@ def build(
+                target=tokens[:, prev_pos + 1 : cur_pos + 1],
+                reduction="none",
+                ignore_index=pad_id,
+            )
+        #####
+
+        if temperature > 0:
+            probs = torch.softmax(logits[:, -1] / temperature, dim=-1) #TODO: eval the perf impact of logits vs. logits[:,-1]
+            next_token = sample_top_p(probs, top_p)
+        else:
+            next_token = torch.argmax(logits[:, -1], dim=-1)
+        next_token = next_token.reshape(-1)
+        # only replace token if prompt has already been generated
+        input_text_mask_tmp = input_text_mask.index_select(1, cur_pos_tensor).squeeze(dim=1)
+        tokens_tmp = tokens.index_select(1, cur_pos_tensor).squeeze(dim=1)
+        next_token = torch.where(
+            input_text_mask_tmp, tokens_tmp, next_token
+        )
+        next_token = next_token.unsqueeze(1)
+        tokens = tokens.index_copy(1, cur_pos_tensor, next_token)
+        input_pos_tensor = input_pos_tensor[-1:] + 1
+        cur_pos_tensor = cur_pos_tensor + 1
+        output_pos_tensor = cur_pos_tensor - 1
+        input_tokens = tokens.index_select(1, input_pos_tensor)
+
+        #TODO: optimize and bring back
+        eos_reached = False
+        #eos_reached |= (~input_text_mask[:, cur_pos]) & (
+        #    next_token == self.tokenizer.eos_id
+        #)
+        return tokens, input_tokens, cur_pos_tensor, input_pos_tensor, output_pos_tensor, cache_kvs, eos_reached
+
+    # @torch.inference_mode() #TORCH_XLA doesn't have this line
     def generate(
         self,
-        prompt_tokens: List[List[int]],
+        #prompt_tokens: List[List[int]], #TODO: LIST OF LIST INSTEAD OF LIST
+        prompts: List[str], #TODO: pass prompt_tokens instead?
         max_gen_len: int,
         temperature: float = 0.6,
         top_p: float = 0.9,
-        logprobs: bool = False,
-        echo: bool = False,
-    ) -> Tuple[List[List[int]], Optional[List[List[float]]]]:
+        logprobs: bool = False,#NEW LINE
+        echo: bool = False,#NEW LINE
+    ) -> Tuple[List[List[int]], Optional[List[List[float]]]]: #NEW OUTPUT FORMAT
         params = self.model.params
-        bsz = len(prompt_tokens)
+        bsz = len(prompts)
+        #bsz = len(prompt_tokens)
         assert bsz <= params.max_batch_size, (bsz, params.max_batch_size)
 
+        prompt_tokens = [self.tokenizer.encode(x, bos=True, eos=False) for x in prompts]
+
+        # TODO: check if we need to optimize this block - llama1 has a simpler logic
         min_prompt_len = min(len(t) for t in prompt_tokens)
         max_prompt_len = max(len(t) for t in prompt_tokens)
         assert max_prompt_len <= params.max_seq_len
         total_len = min(params.max_seq_len, max_gen_len + max_prompt_len)
 
         pad_id = self.tokenizer.pad_id
-        tokens = torch.full((bsz, total_len), pad_id, dtype=torch.long, device="cuda")
+        tokens = torch.full((bsz, total_len), pad_id, dtype=torch.long)
```

Comment:
This was modified in LLaMA 1 to avoid dynamic shape from different inputs.

Reply:
optimized now.